### PR TITLE
[Core] Read resources from an environment variable

### PR DESCRIPTION
**Pull request by wuisawesome**
## Why are these changes needed?

Passing resources as an environment variable is necessary to support multiple autoscaler instance types. 

## Checks

- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/latest/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failure rates at https://ray-travis-tracker.herokuapp.com/.
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested (please justify below)

### CODE REVIEW

File: python/ray/node.py [hunk 1]
```diff
@@ -249,12 +250,25 @@ def _init_temp(self, redis_client):
 
     def get_resource_spec(self):
         """Resolve and return the current resource spec for the node."""
+        resources = {}
+        try:
+            env_string = os.getenv("RAY_RESOURCES")
+            if env_string:
+                resources = json.loads(env_string)
+        except Exception:
```
Comment (rkooo567) [inline python/ray/node.py:258]: Catch json exception instead?

File: python/ray/node.py [hunk 2]
```diff
@@ -249,12 +250,25 @@ def _init_temp(self, redis_client):
...
+            logger.exception(
+                "Unable to parse the RAY_RESOURCES environment variable using "
+                "json.loads, defaulting to empty dict. Try using a"
+                "format like\n\n"
+                "    RAY_RESOURCES='{\"CustomResource1\": 3, "
+                "\"CustomReseource2\": 2}'")
+
         if not self._resource_spec:
+            resources.update(self._ray_params.resources)
```
Comment (rkooo567) [inline python/ray/node.py:267]: Are we overwriting if there are same keys from env vars & given custom resources? Why don't we raise errors in that case?

File: doc/source/configure.rst [hunk 3]
```diff
@@ -50,6 +50,14 @@ If using the command line, connect to the Ray cluster as follow:
   # Connect to ray. Notice if connected to existing cluster, you don't specify resources.
   ray.init(address=<address>)
 
+A subset of Ray options are also available via environment variables.
+
+.. code-block:: python
+    os.environ["RAY_RESOURCES"] = "{\"custom1\":1}"
+    os.environ["RAY_ADDRESS"] = address
+    ray.init()
+
```
Comment (ericl) [inline doc/source/configure.rst:59]: Let's not document these for now.

File: python/ray/node.py [hunk 4]
```diff
@@ -249,12 +250,25 @@ def _init_temp(self, redis_client):
 
     def get_resource_spec(self):
         """Resolve and return the current resource spec for the node."""
+        resources = {}
+        try:
+            env_string = os.getenv("RAY_RESOURCES")
```
Comment (ericl) [inline python/ray/node.py:255]: Please rename this to `RAY_OVERRIDE_RESOURCES`, here and elsewhere.

File: python/ray/node.py (hunk 1 above)
Comment (ericl) [reply python/ray/node.py:258]: Can you actually remove this try block? It's preferable to fail startup if the config is mis-formatted.

File: python/ray/tests/test_advanced_3.py [hunk 5]
```diff
@@ -683,6 +683,16 @@ def test_ray_address_environment_variable(ray_start_cluster):
...
 
 
+def test_ray_resources_environment_variable(ray_start_cluster):
+    address = ray_start_cluster.address
+
+    os.environ["RAY_RESOURCES"] = "{\"custom1\":1}"
+    ray.init(address=address, resources={"custom2": 1})
+
+    assert ("custom1" in ray.cluster_resources()
+            and "custom2" in ray.cluster_resources())
```
Comment (ericl) [inline python/ray/tests/test_advanced_3.py:693]: Can you add a test that the env var has precedence?

**Review by ericl — CHANGES_REQUESTED** [review]

File: python/ray/tests/test_advanced_3.py (hunk 5 above)
Comment (wuisawesome) [reply python/ray/tests/test_advanced_3.py:693]: I thought we decided that env var does **not** have preference over cmd line flag. I'll add a conflict test case either way though.

Comment (ericl) [reply python/ray/tests/test_advanced_3.py:693]: Let's make it override for now, and call the env var name OVERRIDE.

File: python/ray/node.py (hunk 2 above)
Comment (wuisawesome) [reply python/ray/node.py:267]: Now that we're changing this to non-standard behavior we'll emit a warning.

File: python/ray/node.py [hunk 6]
```diff
@@ -249,12 +250,33 @@ def _init_temp(self, redis_client):
 
     def get_resource_spec(self):
         """Resolve and return the current resource spec for the node."""
+
+        def merge_resources(env_dict, params_dict):
+            """Merge two dictionaries, picking from the second in the event of a conflict.
+            Also emmit a warning on every conflict.
```
Comment (ericl) [inline python/ray/node.py:256]:
```suggestion
            Also emit a warning on every conflict.
```

File: python/ray/node.py [hunk 7]
```diff
@@ -249,12 +250,33 @@ def _init_temp(self, redis_client):
...
+            """
+            result = params_dict.copy()
+            result.update(env_dict)
+
+            for key in set(env_dict.keys()).intersection(
+                    set(params_dict.keys())):
+                logger.warning("Autoscaler is overriding your resource:"
```
Comment (ericl) [inline python/ray/node.py:263]:
```suggestion
                logger.warning("RAY_OVERRIDE_RESOURCES is overriding your resource:"
```

**Review by ericl — APPROVED** [review]

**Comment by ericl** [comment]
Looks good but some conflicts.

**Comment by ericl** [comment]
./python/ray/tests/test_advanced_3.py:667:1: W293 blank line contains whitespace

**Review by wuisawesome — COMMENTED** [review]
.

**Pull request merged**